### DISCUSSION TITLE
Subscription behavior

### DISCUSSION THREAD

**Discussion by slore** [FreeOpcUa/opcua-asyncio#652] [Q&A] [answered]
Hello guys,

I am fairly new to opc-ua and I am trying to build an HMI that communicates via opc-ua to the PLC.

In order to update the visualization, after any change in the PLC monitored tags, I am exploiting the subscription strategy instead of polling.

I am seeing a strange behavior in the subscription response time: after a change in the PLC tag it takes even up to 3sec to trigger the event reaction. I have made the same test with UA expert and it's almost instantaneous.
To be honest this response time can vary much: from some millisec to many seconds.

to be more accurate: I am subscriving to 2000 plc tags.

Could you please guide me to correct this behavior?

**Comment by AndreasHeine** [comment, marked as answer]
what publishinterval did you use?

**Reply by slore** [reply]
I have tested both 100 and 500, same exact behavior.

**Reply by AndreasHeine** [reply]
thats not normal!
i hope you dont do strange blocking stuff in the subhandler!?

**Reply by AndreasHeine** [reply]
can you share codesnippets?

**Reply by slore** [reply]
sure.

here it is:

_#Subscription creation
**async def create_subscription(self):**
    handler = SubHandler(self)
    sub = await self.client.create_subscription(500, handler)
    handle = await sub.subscribe_data_change(self.node)_

_#Handler
**class SubHandler(object):**
    
    **def __init__(self, ui_tag: UiTag):**
        self.ui_tag = ui_tag

    **def datachange_notification(self, node, val, data):**
        print(f"Python: New data change event {node}, {val} ts {int(round(time.time() * 1000))}")
        self.ui_tag.update_value(val) #simple value assignation
        self.ui_tag.push_stack() #list append that will be consumed in another task

    **def event_notification(self, event):**
        print("Python: New event", event)_

**Reply by AndreasHeine** [reply]
looks fine... is there any log which says something like -> WARNING:asyncua.client.client:Revised values returned differ from subscription values: CreateSubscriptionResult(SubscriptionId=1094428078, RevisedPublishingInterval=1000.0, RevisedLifetimeCount=300, RevisedMaxKeepAliveCount=30)

```python
logging.basicConfig(level=logging.WARNING)
_logger = logging.getLogger('asyncua')
```

**Reply by slore** [reply]
just tested now: I don't see any of thoose messages in Pycharm console.

**Reply by slore** [reply]
could be this my issue:

"Start the publish loop if it is not yet running
The current strategy is to have only one open publish request per UaClient. This might not be enough
in high latency networks or in case many subscriptions are created. A Set of Tasks of `_publish_loop`
could be used if necessary."

this comment is extracted from : [create_subscription](https://github.com/FreeOpcUa/opcua-asyncio/blob/7d7841bfb7b4e351797b8a5cebdfa68a6418e406/asyncua/client/ua_client.py#L497)

**Reply by AndreasHeine** [reply]
what plc is it? 
and could you make a wireshark trace (i want to see the requests on the wire)?

**Reply by slore** [reply]
it is a Twincat 3 enviroment by Beckhoff that runs on my laptop.
Beckhoff runs PLC on PC so it is the same exact behavior I will have in the real PLC

I will setup wireshark and make a log of it.

**Reply by slore** [reply]
Here the log:
first * the HMI tells to Python to make a opc ua write
second * Python performs the opc ua write
third and fourth * Subscription trigger

I don't understand the line marked with ?

![image](https://user-images.githubusercontent.com/7493781/130611108-9d18456e-e3da-4463-a4f9-c7d285f93d0a.png)

[wireshark log.txt](https://github.com/FreeOpcUa/opcua-asyncio/files/7038857/wireshark.log.txt)

**Reply by AndreasHeine** [reply]
if you use encryption it would be easier to read if it is disabled (Policie -> None) so its possible to look into the requests (e.g. create subscription request ...) and if you export the relevant trace part as .pcapng (you can also filter for "opcua" protocol) and just post that relevant part (.pcapng) here i dont need to guess that much ;)

**Reply by slore** [reply]
I think I have isolated the issue to the Beckhoff opc-ua Server:

I've tested and logged the same behavior also with UA Expert as client.

here the screenshot of the more than 7sec delay between write and change notification:

![image](https://user-images.githubusercontent.com/7493781/130616562-52779d0d-2bd8-426e-acef-11d62cdbaa82.png)

**Reply by AndreasHeine** [reply]
![grafik](https://user-images.githubusercontent.com/56362817/130616941-f2150495-27ab-4eb6-afb2-17fc841e6170.png)

what me wonders is the service faults...

the writerequest has nothing to do with the publishresponse there must be a publishrequest above but the delay may happen because of the service faults!

**Reply by slore** [reply]
sorry, let me explain better:
I modify the value of the tag via UA Expert (so here the write request) and then after 7sec of the delay the value updates also in the visualization of UA Expert

**Reply by AndreasHeine** [reply]
looks odd to me! typically with an unencrypted trace easier to understand (have been trough something similar with siemens plc) the best way to detect the issue is wireshark...

**Reply by slore** [reply]
here the complete log.
You can surelly read it better than me
[wireLogUAExpert.zip](https://github.com/FreeOpcUa/opcua-asyncio/files/7039270/wireLogUAExpert.zip)

**Reply by AndreasHeine** [reply]
![grafik](https://user-images.githubusercontent.com/56362817/130624806-0d47d9da-1630-47ce-af3f-a31611afb4ef.png)

two requests got a timeout...

**Reply by slore** [reply]
What still is unclear to me is: is the server not responding or the client not catching the message?

**Reply by AndreasHeine** [reply]
the server sends the servicefault and you should get a exception on clientside aswell in the log

**Reply by slore** [reply]
since last time, I have made tons of test:
it looks like that the number of subscriptions make the system overloaded. Is there any chance to improve client side the performance to handle such a big number of subscriptions?

**Reply by AndreasHeine** [reply]
how many subs do you have?

> 1 Sub with 2k MonitoredItems 
> or
> 2k Subs with 1 MonitoredItem

**Reply by AndreasHeine** [reply]
typically the first is recommended (1 Sub with 2k MonitoredItems)

**Reply by slore** [reply]
Here could be the issue: I have 2kSubs with 1MonitoredItem.

How can I do the other way around?

Thanks!!

**Reply by AndreasHeine** [reply]
first of all some background: https://reference.opcfoundation.org/v104/Core/docs/Part4/5.13.1/
create one sub and with that sub call `await sub.subscribe_data_change(nodelist)` (nodelist = [...2k nodes....]) and inside the SubHandler you can make sorting to a python variable (according to the node-id) or putting it just in a queue in order to process it in a different task or subprocess (the important thing is do not do blocking stuff in the SubHandler the handler method will be called for each datachange notification in the subscriptions publishresponse)

**Reply by slore** [reply]
I want to thank you, I defenitelly solved my issue with your help.
